File: example_code/Optimnization&Vectorization/Exercise/C/profiling.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#define N 1000
#define ITERMAX 100

double Vecsum(double X[][N],int j);
double Vecprod(double X[][N],int j);

double WallTime()
{
    struct timespec tp;
    clock_gettime(CLOCK_REALTIME,&tp);
    return ((double)tp.tv_sec + (double)tp.tv_nsec * 1.e-9);
}


int main(void)
{
  int i,j,it;
  double A[N], B[N][N],C[N][N],D[N][N];
  double stime,etime;
  double dtime,flops;

  for(i=0;i<N;i++){
    for(j=0;j<N;j++){
      B[i][j]=(double)(rand())/RAND_MAX;
      C[i][j]=(double)(rand())/RAND_MAX;
      D[i][j]=(double)(rand())/RAND_MAX;
    }
  }

  stime=WallTime();
  for(it=0;it<ITERMAX;it++){
    for(j=0;j<N;j++){
      A[j]=Vecsum(B,j)+Vecprod(C,j)*Vecsum(D,j);
    }
    if(it%100==0) printf("%f, %f, %f, %f\n",A[0],B[0][0],C[0][0],D[0][0]);
  }
  etime=WallTime();
  dtime=etime-stime;
  flops=(1.0e-6*(ITERMAX)*N*(3.0*N+2))/dtime;
  printf("loop time = %12.5f seconds\n",dtime);
  printf("loops runs at %15.5f MFLOPS\n",flops);
  return 0;
}

double Vecsum(double X[][N],int j)
{
  int i;
  double tmp;
  tmp=X[0][j];
  for(i=1;i<N;i++)
    tmp = tmp+X[j][i];
  return tmp;
}

double Vecprod(double X[][N],int j)
{
  int i;
  double tmp;
  tmp=X[0][j];
  for(i=1;i<N;i++)
    tmp=tmp+X[j][i];
  return tmp;
}

File: example_code/Optimnization&Vectorization/Solution/Fortran/profiling.f90
PROGRAM PROFILING
IMPLICIT NONE
INTEGER,PARAMETER::N=1000,ITERMAX=100
INTEGER::IT,I,J
REAL,DIMENSION(N,N)::B,C,D
REAL,DIMENSION(N)::A
INTEGER(KIND=8)::ST,ET,C_RATE
REAL(KIND=8)::DTIME, FLOPS

CALL RANDOM_NUMBER(B)
CALL RANDOM_NUMBER(C)
CALL RANDOM_NUMBER(D)

CALL SYSTEM_CLOCK(ST,C_RATE)
DO IT=1,ITERMAX
  DO J=1,N
    A(J)=VECSUM(B,J)+VECPROD(C,J)*VECSUM(D,J)
  END DO
  IF(MOD(IT,100)==0) WRITE(*,*) A(1),B(1,1),C(1,1),D(1,1)
END DO
CALL SYSTEM_CLOCK(ET)
DTIME=(ET-ST)/DBLE(C_RATE)
FLOPS=(1.0E-6*ITERMAX*N*(3*N+2))/DTIME
WRITE(*,100) DTIME
WRITE(*,200) FLOPS

100 FORMAT("loop time = ",f12.5," seconds")
200 FORMAT("loop runs at ",f15.5," MFLOPS")

CONTAINS
REAL FUNCTION VECSUM(X,J)
IMPLICIT NONE
REAL,DIMENSION(N,N),INTENT(IN)::X
INTEGER,INTENT(IN)::J
INTEGER::I
VECSUM=X(1,J)
DO I=2,N
  VECSUM=VECSUM+X(I,J)
END DO
END FUNCTION VECSUM

REAL FUNCTION VECPROD(X,J)
IMPLICIT NONE
REAL,DIMENSION(N,N),INTENT(IN)::X
INTEGER,INTENT(IN)::J
INTEGER::I
VECPROD=X(1,J)
DO I=2,N
  VECPROD=VECPROD*X(I,J)
END DO
END FUNCTION VECPROD

END PROGRAM PROFILING